pinecone/control/pinecone.py

```diff
@@ -0,0 +1,277 @@
+                index_type=index_type,
+                metric=metric,
+                replicas=replicas,
+                shards=shards,
+                pods=pods,
+                pod_type=pod_type,
+                index_config=index_config or {},
+                metadata_config=metadata_config,
+                source_collection=source_collection,
+            )
+        )
+
+        def is_ready():
+            status = self._get_status(name)
+            ready = status["ready"]
+            return ready
+
+        if timeout == -1:
+            return
+        if timeout is None:
+            while not is_ready():
+                time.sleep(5)
+        else:
+            while (not is_ready()) and timeout >= 0:
+                time.sleep(5)
+                timeout -= 5
+        if timeout and timeout < 0:
+            raise (
+                TimeoutError(
+                    "Please call the describe_index API ({}) to confirm index status.".format(
+                        "https://www.pinecone.io/docs/api/operation/describe_index/"
+                    )
+                )
+            )
+
+    def delete_index(self, name: str, timeout: int = None):
+        """Deletes a Pinecone index.
+
+        :param name: the name of the index.
+        :type name: str
+        :param timeout: Timeout for wait until index gets ready. If None, wait indefinitely; if >=0, time out after this many seconds;
+            if -1, return immediately and do not wait. Default: None
+        :type timeout: int, optional
+        """
+        api_instance = self.index_api
+        api_instance.delete_index(name)
+        self.index_host_store.delete_host(self.config, name)
+
+        def get_remaining():
+            return name in api_instance.list_indexes()
+
+        if timeout == -1:
+            return
+
+        if timeout is None:
+            while get_remaining():
+                time.sleep(5)
+        else:
+            while get_remaining() and timeout >= 0:
+                time.sleep(5)
+                timeout -= 5
+        if timeout and timeout < 0:
+            raise (
+                TimeoutError(
+                    "Please call the list_indexes API ({}) to confirm if index is deleted".format(
+                        "https://www.pinecone.io/docs/api/operation/list_indexes/"
+                    )
+                )
+            )
+
+    def list_indexes(self):
+        """Lists all indexes."""
+        response = self.index_api.list_indexes()
+        return response
+
+    def describe_index(self, name: str):
+        """Describes a Pinecone index.
+
+        :param name: the name of the index to describe.
+        :return: Returns an `IndexDescription` object
+        """
+        api_instance = self.index_api
+        response = api_instance.describe_index(name)
+        db = response.database
+        host = response.status.host
+
+        self.index_host_store.set_host(self.config, name, "https://" + host)
```

Comment:
We proactively stick this host url in the singleton store if they happen to call describe_index before doing any index operations. Just to save an additional round trip.